unittests/BuildSystem/BuildSystemFrontendTest.cpp

```diff
@@ -316,6 +317,122 @@ commandFinished: 3: 0
 
 }
 
+
+TEST_F(BuildSystemFrontendTest, missingShellArguments) {
+  writeBuildFile(R"END(
+client:
+    name: basic
+
+commands:
+    C0:
+        tool: shell
+        args: []
+)END");
+
+  TestBuildSystemFrontendDelegate delegate(sourceMgr);
+
+  BuildSystemFrontend frontend(delegate, invocation, createLocalFileSystem());
+  ASSERT_FALSE(frontend.build(""));
+
+  ASSERT_TRUE(delegate.checkTrace(R"END(
+error: unable to configure client
+error: unable to load build file
+)END"));
+}
+
+
+
+class TestCancellationFrontendDelegate : public TestBuildSystemFrontendDelegate {
+  using super = TestBuildSystemFrontendDelegate;
+
+  std::mutex shouldCancelMutex;
+  std::condition_variable shouldCancelCond;
+  std::atomic<bool> shouldCancel{false};
+  std::atomic<bool> done{false};
+
+  std::thread cancelWorker;
+
+public:
+  TestCancellationFrontendDelegate(SourceMgr& sourceMgr)
+  : TestBuildSystemFrontendDelegate(sourceMgr),
+    cancelWorker(&TestCancellationFrontendDelegate::cancelHandler, this)
+  {
+  }
+
+  ~TestCancellationFrontendDelegate() {
+    done = true;
+    shouldCancelCond.notify_all();
+    cancelWorker.join();
+  }
+
+  std::atomic<bool> cancelAtStart{false};
+  std::atomic<bool> cancelAtCommand{false};
+
+  virtual void commandPreparing(Command* command) override {
+    super::commandPreparing(command);
+
+    if (cancelAtStart) {
+      shouldCancel = true;
+      shouldCancelCond.notify_all();
+    }
+  }
+
+  virtual void commandProcessStarted(Command* command, ProcessHandle handle) override {
+    super::commandProcessStarted(command, handle);
+
+    if (cancelAtCommand) {
+      shouldCancel = true;
+      shouldCancelCond.notify_all();
+    }
+  }
+
+private:
+  void cancelHandler() {
+    std::unique_lock<std::mutex> lock(shouldCancelMutex);
+    while (!done) {
+      if (shouldCancel) {
+        cancel();
+        shouldCancel = false;
+      }
+      shouldCancelCond.wait(lock);
+    }
+  }
+};
+
+
+TEST_F(BuildSystemFrontendTest, cancellation) {
+  writeBuildFile(R"END(
+client:
+    name: client
+
+targets:
+    "": ["output"]
+
+commands:
+    C0:
+        tool: shell
+        outputs: ["output"]
+        args: yes >/dev/null
+)END");
+
+  TestCancellationFrontendDelegate delegate(sourceMgr);
+  BuildSystemFrontend frontend(delegate, invocation, createLocalFileSystem());
+
+  for (int i = 0; i < 100; ++i) {
```

Comment:
I have set this iteration count to a reasonable size for ongoing checking.   However, I have run it for much larger numbers to qualify this change.   I have also run more than 200k parallel build/cancel tests driven via an external client.

Reply:
Nice. Maybe drop a FIXME here saying we could up if it we had a “long running tests” mode?